{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "096130a2d7cd8ec859f14bf7305e3047",
     "grade": false,
     "grade_id": "cell-5cd566ae8d583383",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# COMP90016 - Assignment 2\n",
    "Version 1 Last edited 10/4/2024\n",
    "\n",
    "## Semester 1, 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NAME = \"Keziah Tikno\"\n",
    "\n",
    "ID = \"1319716\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "287a2590ecc61503da17000ed2e6e857",
     "grade": false,
     "grade_id": "cell-ea0e1a6df37e43f5",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "This assignment should be completed by each student individually. Make sure you read this entire document, and ask for help if anything is not clear. Any changes or clarifications to this document will be announced via the LMS.\n",
    "\n",
    "Please make sure you review the University's rules on academic integrity: https://academicintegrity.unimelb.edu.au/\n",
    "\n",
    "You submission must be your own work. Do not copy material from other students, from the internet or from AI tools. \n",
    "\n",
    "Your completed notebook file containing all your answers will be turned in via Canvas. Please also submit an HTML file with the output cleared.\n",
    "\n",
    "To complete the assignment, finish the tasks in this notebook.\n",
    "\n",
    "The tasks are a combination of writing your own code, interpreting the results and answering related short-answer questions.\n",
    "\n",
    "In some cases, we have provided test input and test output that you can use to try out your solutions. These tests are just samples and are **not** exhaustive - they may warn you if you've made a mistake, but they are not guaranteed to. It's up to you to decide whether your code is correct.\n",
    "\n",
    "**Remember to save your work early and often.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "4303165c210b02050c183fa566e20543",
     "grade": false,
     "grade_id": "cell-a654a22c19fdefbe",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Marking\n",
    "\n",
    "Cells that must be completed to receive marks are clearly labelled. Some cells are code cells, in which you must complete the code to solve a problem. Others are markdown cells, in which you must write your answers to short-answer questions. \n",
    "\n",
    "Cells that must be completed to receive marks are labelled like this:\n",
    "\n",
    "`# -- GRADED CELL (1 mark) - complete this cell --`\n",
    "\n",
    "Some graded cells are code cells, in which you must complete the code to solve a problem. Other graded cells are markdown cells, in which you must write your answers to short-answer questions. \n",
    "\n",
    "You will see the following text in graded code cells:\n",
    "\n",
    "```\n",
    "# YOUR CODE HERE\n",
    "raise NotImplementedError()\n",
    "```\n",
    "\n",
    "***You must remove the `raise NotImplementedError()` line from the cell, and replace it with your solution.***\n",
    "\n",
    "Only add answers to graded cells. If you want to import a library or use a helper function, this must be included in a graded cell.\n",
    "\n",
    "Only graded cells will be marked.\n",
    "**Don't make changes outside graded cells, and don't add or remove cells from the notebook**.\n",
    "\n",
    ">Word limits, where stated, will be strictly enforced. Answers exceeding the limit **will not be marked**.\n",
    "\n",
    ">Run-time limits will be imposed for each coding question. The run-time of a code cell can be calculated by including `%time` at the top of your cell. Cells exceeding the run-time limit **will not be marked**. The run-time limits only apply to test cases that are included in this document.\n",
    "\n",
    "No marks are allocated to commenting in your code. We do however, encourage efficient and well-commented code.\n",
    "\n",
    "The total marks for the assignment add up to 100, and it will be worth 15% of your overall subject grade.\n",
    "\n",
    "Part 1: 35 marks\n",
    "\n",
    "Part 2: 20 marks\n",
    "\n",
    "Part 3: 45 marks\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "3aaed7e89af84c07ff25bb32364b91b1",
     "grade": false,
     "grade_id": "cell-a8f3f9ddb8997b46",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Submitting\n",
    "\n",
    "Before you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\\rightarrow$Run All).\n",
    "\n",
    "Make sure you fill in any place that says `YOUR CODE HERE` or \"YOUR ANSWER HERE\", as well as your name and student ID at the top of this notebook.\n",
    "\n",
    "\n",
    "Your completed notebook file containing all your answers must be turned in via LMS in `.ipynb` format.\n",
    "You must also submit a copy of this notebook in `html` format with the output cleared.\n",
    "You can do this by using the `clear all output` option in the menu.\n",
    "\n",
    "Your submission should include **only two** files with names formatted as: **Assignment_2.ipynb** and **Assignment_2.html**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "65f5920d6e302563ed6f4b8c00c727af",
     "grade": false,
     "grade_id": "cell-231ce2cdcc452d8d",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Overview\n",
    "\n",
    "In this assignment, you will answer questions about VCF files and BLAST.\n",
    "\n",
    "You will use the `pysam` and `biopython` libraries in your functions. You may want to refer to sections of the documentation for these tools for additional help. Additional to `pysam`, `biopython` and standard Python 3 functions and methods, you may also use any other library we have used in Computational Genomics including `collections`, `numpy`, `pandas`, `math`, `itertools`, `seaborn` and `matplotlib`.\n",
    "\n",
    "You will also use the `BLAST` command-line tool. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "b1e9aace2837e55c85c9be2b30a0c389",
     "grade": false,
     "grade_id": "cell-cb61bd3dd20d0831",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "5d20151ec3cc0e20534126b58fd0b79a",
     "grade": false,
     "grade_id": "cell-eca7f6df2f0e77fc",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tiknokeziah/Desktop/Keziah/School/COMP90016/venv/lib/python3.9/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import requests\n",
    "from IPython.core.display import HTML\n",
    "\n",
    "# Function to get data. DO NOT MODIFY!\n",
    "def fetch_file(url, outpath='.'):\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        print('File found!')\n",
    "        # Get the filename from the URL\n",
    "        filename = os.path.basename(url).split('?', 1)[0]\n",
    "        # Construct the filepath using the specified directory and filename\n",
    "        filepath = os.path.join(outpath, filename)\n",
    "        # Create the directory if it doesn't exist\n",
    "        if not os.path.exists(outpath):\n",
    "            print(f'Creating output dir: {outpath}')\n",
    "            os.makedirs(outpath)\n",
    "        # Check if the file already exists in the specified directory\n",
    "        if os.path.exists(filepath):\n",
    "            print(f'{filename} already exists in {outpath}. Skip download.')\n",
    "        else:\n",
    "            with open(filepath, 'wb') as f:\n",
    "                f.write(response.content)\n",
    "                f.close()\n",
    "            print(f'Saved to: {filepath}')\n",
    "    else:\n",
    "        print(f'File not found: Code {response.status_code}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>h1 {\n",
       "    border: 3.0px solid #333;\n",
       "    padding: 12px 18px;\n",
       "    background-color: #C8D3FD;\n",
       "}\n",
       "\n",
       "h2 {\n",
       "    padding: 12px 18px;\n",
       "    background-color: #C8D3FD;\n",
       "}\n",
       "\n",
       ".question {\n",
       "    color: rgb(0, 96, 100);\n",
       "    background: rgb(178, 235, 242);\n",
       "    border: solid 1px rgb(77, 208, 225);\n",
       "    padding: 10px;\n",
       "}\n",
       "\n",
       ".info {\n",
       "    color: rgb(27, 94, 32);\n",
       "    background: rgb(200, 230, 201);\n",
       "    border: solid 1px rgb(129, 199, 132);\n",
       "    padding: 10px;\n",
       "}\n",
       "\n",
       "</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make the notebook pretty\n",
    "HTML(requests.get('https://raw.githubusercontent.com/melbournebioinformatics/COMP90016/main/data/2023/style/custom.css').text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "93b5a3ed48b832e8331a2c450644b810",
     "grade": false,
     "grade_id": "cell-0664141ba25828c7",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File found!\n",
      "comp90016_assignment_2_demo_a.vcf already exists in data. Skip download.\n",
      "File found!\n",
      "comp90016_assignment_2_demo_b.vcf already exists in data. Skip download.\n",
      "File found!\n",
      "comp90016_assignment_2.vcf already exists in data. Skip download.\n",
      "File found!\n",
      "comp90016_assignment_2_mbp.vcf already exists in data. Skip download.\n",
      "File found!\n",
      "comp90016_assignment_2_COVID19_N.fasta already exists in data. Skip download.\n",
      "File found!\n",
      "comp90016_assignment_2_9b.fasta already exists in data. Skip download.\n"
     ]
    }
   ],
   "source": [
    "# Fetch assignment data\n",
    "f_names = [\"comp90016_assignment_2_demo_a.vcf\",\n",
    "           \"comp90016_assignment_2_demo_b.vcf\",\n",
    "           \"comp90016_assignment_2.vcf\",\n",
    "           \"comp90016_assignment_2_mbp.vcf\",\n",
    "           \"comp90016_assignment_2_COVID19_N.fasta\",\n",
    "           \"comp90016_assignment_2_9b.fasta\"]\n",
    "\n",
    "for filename in f_names:\n",
    "    url = f'https://github.com/melbournebioinformatics/COMP90016/blob/main/data/2024/Assignment_02/data/{filename}?raw=true'\n",
    "    fetch_file(url,'data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "e156ac431aa3baa15f26452429c8b470",
     "grade": false,
     "grade_id": "cell-364cd6139bcc3f75",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Part 1: VCF files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "4a05cdcdf36ac45d312a53bb91b4c3cb",
     "grade": false,
     "grade_id": "cell-682a0639c3e0822b",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "447612e420e0bc606a3e6c83070fcbc9",
     "grade": false,
     "grade_id": "cell-fef38625be02d52f",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import the pysam module.\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "4c428340d4adbee156e79d9f2eb86868",
     "grade": false,
     "grade_id": "cell-44939d002f776371",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the VCF files to produce VariantFile objects.\n",
    "demo_vcf_a = pysam.VariantFile(\"data/comp90016_assignment_2_demo_a.vcf\")\n",
    "demo_vcf_b = pysam.VariantFile(\"data/comp90016_assignment_2_demo_b.vcf\")\n",
    "comp90016_vcf = pysam.VariantFile(\"data/comp90016_assignment_2.vcf\")\n",
    "\n",
    "#Confirm object type\n",
    "comp90016_vcf.is_vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "377eaab19014e23b29c3fb833fec1ade",
     "grade": false,
     "grade_id": "cell-265f6268c5666b8b",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "VariantFile objects allow us to access the data within VCF files. You may want to refer back to the work you did in workshop 5. You may also want to read the documentation linked below:\n",
    ">https://pysam.readthedocs.io/en/latest/usage.html#working-with-vcf-bcf-formatted-files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "e54a6a5b02628456623c70171a7111a6",
     "grade": false,
     "grade_id": "cell-76e121e33e817783",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "### Questions\n",
    "In the cells below, complete the following tasks:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "815f90d71d1d963e7f370312008e52b7",
     "grade": false,
     "grade_id": "Q-1-1",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "<h3>Question 1.1</h3>\n",
    "    \n",
    "(5 marks, max 50 words)\n",
    "\n",
    "Which reference genome (full species name and version) was used for the variant calling  in `comp90016_assignment_2.vcf`? Which variant calling tool was used?\n",
    "\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "Full species = Homo sapiens, version = hg38\n",
    "\n",
    "Variant calling tool = freeBayes v1.3.1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "197807c9fc542d4cb2aa7391c10366c3",
     "grade": false,
     "grade_id": "Q-1-2-info",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "<div class=\"info\">\n",
    "<h3> Question 1.2 </h3>\n",
    "\n",
    "(10 marks)  \n",
    "    \n",
    "The four standard DNA bases can be divided into purines (A, G) and pyrimidines (C,T). A transition is a single base change from a purine to a purine, or from a pyrimidine to a pyrimidine. A transversion is a single base change from a purine to a pyrimidine or from a pyrimidine to a purine. The transition/transversion ratio (ti/tv) is the ratio of the number of transitions to the number of transversions.\n",
    " \n",
    "<b>Challenge:</b> Write a Python function to calculate the ti/tv for a set of SNVs in a VCF file containing variants from a single sample.\n",
    "\n",
    "- [ ] Input: a pysam VariantFile object.\n",
    "- [ ] Output: Return a single floating-point number. Round to 2 decimal places.\n",
    "- [ ] Disregard variants that are not SNVs.\n",
    "- [ ] Consider multi-allelic SNVs as seperate SNVs.\n",
    "- [ ] Consider a particular SNV allele only once, regardless of how many copies of the allele are present.\n",
    "- [ ] If there are no transversion SNVs, return None.\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "5ed0d5c0c9cfadfa552cd7c46945eb4c",
     "grade": false,
     "grade_id": "Q-1-2",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# GRADED CELL 1.2 (10 marks, max 1 min run-time)\n",
    "%time\n",
    "def titv(vcf):\n",
    "    \"\"\"\n",
    "    Write a Python function to calculate the ti/tv for a set of SNVs in a VCF file containing variants from a single sample.\n",
    "    Input: a pysam VariantFile object.\n",
    "    Output: Return a single floating-point number.\n",
    "    Disregard variants that are not SNVs.\n",
    "    Consider multi-allelic SNVs as seperate SNVs.\n",
    "    Consider a particular SNV allele only once, regardless of how many copies of the allele are present.\n",
    "    If there are no transversion SNVs, return None.\n",
    "    \"\"\"\n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "    type_dict = {\"A\": \"purines\", \"G\": \"purines\", \"C\": \"pyrimidines\", \"T\": \"pyrimidines\"}\n",
    "    snv_dict = {}\n",
    "    \n",
    "    ti=0\n",
    "    tv=0\n",
    "    \n",
    "    for variant in vcf.fetch():\n",
    "        ref = variant.ref\n",
    "        alt = variant.alts[0]\n",
    "        \n",
    "        if len(ref) == 1 and len(alt)== 1:\n",
    "            snv_dict[variant.pos] = {}\n",
    "            \n",
    "\n",
    "            if ref not in snv_dict[variant.pos]:\n",
    "                snv_dict[variant.pos][ref] = alt\n",
    "                \n",
    "                if type_dict[ref] == type_dict[alt]:\n",
    "                    ti+=1\n",
    "                elif type_dict[ref] != type_dict[alt]:\n",
    "                    tv+=1\n",
    "\n",
    "    if tv != 0:\n",
    "        ratio = ti/tv\n",
    "    else:\n",
    "        ratio = None\n",
    "    \n",
    "    return round(ratio, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "None\n",
      "A\n",
      "('<CN0>',)\n",
      "<CN0>\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "None\n",
      "GGACAGACAGACAGACAG\n",
      "('GGACAGACAGACAG',)\n",
      "GGACAGACAGACAG\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "CGCACGTCCT\n",
      "('CGCCCGCCCC', 'CGCACGCCCC')\n",
      "CGCCCGCCCC\n",
      "0\n",
      "0\n",
      "CGCACGCCCC\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "CAGTGGGT\n",
      "('GAGGGGGG', 'CAGGGGGG', 'CAGTGGGG')\n",
      "GAGGGGGG\n",
      "0\n",
      "0\n",
      "CAGGGGGG\n",
      "0\n",
      "0\n",
      "CAGTGGGG\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "TCCA\n",
      "('CCCC', 'TCCC')\n",
      "CCCC\n",
      "0\n",
      "0\n",
      "TCCC\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "TCCT\n",
      "('CCCC', 'TCCC')\n",
      "CCCC\n",
      "0\n",
      "0\n",
      "TCCC\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "ACCAA\n",
      "('CCCAC',)\n",
      "CCCAC\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "GCAG\n",
      "('CCAC', 'GCAC')\n",
      "CCAC\n",
      "0\n",
      "0\n",
      "GCAC\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "AGGA\n",
      "('GGGG',)\n",
      "GGGG\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "TCG\n",
      "('TCC',)\n",
      "TCC\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "GTCT\n",
      "('GT',)\n",
      "GT\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "CAG\n",
      "('CG',)\n",
      "CG\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "GAAAAAAAAAAAAAAAAACAAGAAAGATAAAAGAAAACTG\n",
      "('GAAAAAAAAAAAAAAAACAAGAAAGATAAAAGAAAACTG',)\n",
      "GAAAAAAAAAAAAAAAACAAGAAAGATAAAAGAAAACTG\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "AC\n",
      "('GT',)\n",
      "GT\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "GTTTTTTTTTTTTGCTTTAC\n",
      "('GTTTTTTTTTTTTTGCTTTAC',)\n",
      "GTTTTTTTTTTTTTGCTTTAC\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('T',)\n",
      "T\n",
      "True\n",
      "0\n",
      "0\n",
      "G\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "A\n",
      "('G',)\n",
      "G\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "T\n",
      "('C',)\n",
      "C\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "C\n",
      "('A',)\n",
      "A\n",
      "True\n",
      "0\n",
      "0\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Test your function in this cell\n",
    "print(titv(demo_vcf_a)) # should output 2.0\n",
    "print(titv(demo_vcf_b)) # should output 1.0\n",
    "\n",
    "print(titv(comp90016_vcf))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "5cf9739c48851ff23ea3cf51066f0cf2",
     "grade": true,
     "grade_id": "T-1-2",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# --- AUTOGRADING CELL DO NOT EDIT ----\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "ffe6e33589debfc9be1184de6e7933f8",
     "grade": false,
     "grade_id": "Q-1-3",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "<h3>Question 1.3</h3>\n",
    "    \n",
    "(10 marks, max 100 words)\n",
    "\n",
    "Explain why the ti/tv ratio is typically higher in protein-coding regions than in non-coding regions.\n",
    "\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "ti/tv ratio high means there are more transitions than transversions. Transitions are favoured in evolution due to their minimal impact on the DNA structure and their tendency to mutate less disruptively on protein function, especially when occurring synonymously. Although nonsynonymous transitions exist, its alterations in amino acid physicochemical properties are less deleterious compared to nonsynonymous transversions. Due to these reasons, transitions are strongly preferred in conserved protein-coding regions where preserving functionality is essential. Additionally, non-coding regions, experiencing lower selective pressure, may have fewer constraints against transversions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "66ce96c8257cd7846668c011f616ba3f",
     "grade": false,
     "grade_id": "Q-1-4",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "<h3>Question 1.4</h3>\n",
    "    \n",
    "(5 marks, max 100 words)\n",
    "\n",
    "When working on a variant calling project, you might generate files with the extensions vcf.gz or vcf.tbi. Explain the contents of these files and how they relate to .vcf files.\n",
    "\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "vcf.gz is a compressed version of a VCF file, which reduces file size and saves storage space. It has the same information as the uncompressed .vcf file.\n",
    "\n",
    "vcf.tbi is an index file associated with VCF file. The file contains genomic coordinates (chromosome, start and end positions), binning information (organize genomic regions into bins for a fast and efficient data retrieval), index file, reference sequence metadata etc. With these features, it provides efficient access to genomic regions in the VCF file, rapid querying and retrieval of variants within certain genomic intervals."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "961e95eee4edc4987382b2c95fcecca2",
     "grade": false,
     "grade_id": "Q-1-5",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "<h3>Question 1.5</h3>\n",
    "    \n",
    "(5 marks, max 100 words)\n",
    "\n",
    "Computational genomics often involves working with large data sets. Explain how .gz and .tbi files help make working with large files feasible.\n",
    "\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "Large genomic data sets contain thousands to millions of variants which can occupy storage space. Therefore, compressing using .gz reduces the file size 2x-10x while making it more manageable for storage and data transfer. For .tbi file it plays an important role in accessing specific regions within large genomic data. Since, it indexed the genomic regions into bins, which is like a dictionary, the tool just need to look the indexing up and find particular regions within the compressed VCF file. It makes an efficient retrieval of variants without having to decompress and read the entire file sequentially."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "883a9e149b5ae95d85c9430e3e80184d",
     "grade": false,
     "grade_id": "cell-364cd6139bcc3f76",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Part 2: More VCF files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "287a7568ccafdc31a916f733eea2d857",
     "grade": false,
     "grade_id": "cell-682a0639c3e0822c",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "66f838b8486e1693094eab0812821fcc",
     "grade": false,
     "grade_id": "cell-e05d22a28e146b02",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "The data you will be using are publically available as part of the 1000 Genomes Project, which was completed in 2015. It was one of the first large-scale projects to try and capture the level of variation in the human genome. You can read one of the key publiciations describing this dataset here: https://www.nature.com/articles/nature15393. \n",
    "\n",
    "Our gene of interest is *MBP* (Myelin Basic Protein, NM_001025081). Variants in *MBP* from three selected individuals (HG02687, HG03897 and HG04118) have been included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "3b7759e2155dcf04b66356506381a88b",
     "grade": false,
     "grade_id": "cell-be8261b79d5046a3",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# Read in the MBP VCF file to produce a VariantFile objects.\n",
    "comp90016_vcf_MBP = pysam.VariantFile(\"data/comp90016_assignment_2_mbp.vcf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "2ddc7e4f5f2c1e95f23e8867375c091d",
     "grade": false,
     "grade_id": "cell-76e121e33e817784",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "### Questions\n",
    "In the cells below, complete the following tasks:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "4f3e4e8fcec00b8bca9305e2e96d87dc",
     "grade": false,
     "grade_id": "Q-2-1-info",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "<div class=\"info\">\n",
    "<h3> Question 2.1 </h3>\n",
    "\n",
    "(10 marks)  \n",
    " \n",
    "<b>Challenge:</b> Write a Python function to create a dictionary of for a given variant position across all samples in a pysam VariantFile object.\n",
    "\n",
    "- [ ] Input:\n",
    "    - a pysam VariantFile object (vcf)\n",
    "    - the name of a contig as a string (chrom)\n",
    "    - an integer describing a position (pos)\n",
    "- [ ] Output: A dictionary of genotypes with sample names as keys and tuples of the genotypes as values.\n",
    "- [ ] If the combination of chrom and pos does not exist in vcf, return None.\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "d1185e0ce8dfe1d1346793042d0947e3",
     "grade": false,
     "grade_id": "Q-2-1",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# GRADED CELL 2.1 (10 marks, max 1 min run-time)\n",
    "%time\n",
    "def genotype_dict(vcf, chrom, pos):\n",
    "    \"\"\"\n",
    "    Create a genotype dictionary for a given variant position in a pysam VariantFile object. \n",
    "    The sample names are keys and tuples of the genotypes are the values. \n",
    "    Assume vcf is a pysam VariantFile object. \n",
    "    Assume chrom is a string. \n",
    "    Assume pos is an integer. \n",
    "    If the combination of chrom and pos does not exist in vcf, return None. \n",
    "    Return the genotype dictionary.\n",
    "    \"\"\"\n",
    "    \n",
    "    # YOUR CODE HERE\n",
    "    geno_dict = {}\n",
    "    \n",
    "    for variant in vcf.fetch():\n",
    "        \n",
    "        if variant.chrom == chrom and variant.pos == pos:\n",
    "            \n",
    "            for geno in vcf.header.samples:\n",
    "                geno_dict[geno] = variant.samples[geno].values()[0]\n",
    "\n",
    "    if len(geno_dict) == 0:\n",
    "        return None\n",
    "        \n",
    "    return geno_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test your function in this cell\n",
    "print(genotype_dict(demo_vcf_b, '18', 74690979)) # should output {'HG02687': (0, 0), 'HG03897': (0, 1), 'HG04118': (0, 0)}\n",
    "print(genotype_dict(demo_vcf_b, '2', 1000)) # should output None\n",
    "\n",
    "print(genotype_dict(comp90016_vcf_MBP, '18', 74694135))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "23884c6c11f6a2b89ef06df7d4692b8f",
     "grade": true,
     "grade_id": "T-2-1",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# --- AUTOGRADING CELL DO NOT EDIT ----\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "4c5629594e0648b98ccffe4ebaf22ab7",
     "grade": false,
     "grade_id": "Q-2-2",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "<h3>Question 2.2</h3>\n",
    "    \n",
    "(5 marks, max 100 words)\n",
    "\n",
    "For each sample, list the total number of homozygous reference, homozygous alternative and heterozygous alleles. Which of the three samples has the most homozygous alternative variants in the MBP gene?\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "counts_dict = {}\n",
    "inter_count = {}\n",
    "\n",
    "for i in comp90016_vcf_MBP.fetch():\n",
    "\n",
    "    for j in i.samples:\n",
    "        if j not in inter_count:\n",
    "            inter_count[j] = [i.samples[j]['GT']]\n",
    "            \n",
    "        else:\n",
    "            inter_count[j].append(i.samples[j]['GT'])\n",
    "\n",
    "for key in inter_count:\n",
    "    counts_dict[key] = dict(Counter(inter_count[key]))\n",
    "\n",
    "print(counts_dict)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "HG02687: 0|0 (homo. ref) = 1317, 0|1 or 1|0 (hetero.) = 68, 1|1 (homo. alt) = 22\n",
    "HG03897: 0|0 (homo. ref) = 1349, 0|1 or 1|0 (hetero.) = 42, 1|1 (homo. alt) = 16\n",
    "HG04118: 0|0 (homo. ref) = 1331, 0|1 or 1|0 (hetero.) = 58, 1|1 (homo. alt) = 18"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "e5c164cab8c9b17020b5f60cb8d3bd3d",
     "grade": false,
     "grade_id": "Q-2-3",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "<h3>Question 2.3</h3>\n",
    "    \n",
    "(5 marks, max 100 words)\n",
    "\n",
    "What information is contained in the ID column of the MBP VCF file? Explain why this information is useful.\n",
    "    \n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "It is a unique label to identify a particular SNVs. They serve as crucial variant identifiers for researchers and databases, offering a standardized naming convention for most SNPs. These IDs are particularly valuable as they facilitate variant annotation and linking to databases like dbSNP (started with 'rs'). This linkage enables researchers to gain deeper insights into variants, including their association with traits, population genetic patterns, and implications for clinical interpretation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "f325258f0616548a3170a83e2d91976a",
     "grade": false,
     "grade_id": "cell-364cd6139bcc3f74",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Part 3: BLAST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "7db14284183089f6564b2e55eb5852a6",
     "grade": false,
     "grade_id": "cell-7bb9f56fdeefa80e",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "d556e0aa3cd0c13a38b45f19e159166d",
     "grade": false,
     "grade_id": "cell-9ed8653a6f115a66",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "The gene ORF-9b has been detected in multiple isolates of SARS coronavirus as well as in other coronaviruses that infect non-human animals. We will determine whether SARS-CoV-2 (responsible for COVID-19) has an ORF that is homologous to ORF-9b in other coronaviruses. We will be using `BLAST` on the command line for this purpose.\n",
    "\n",
    "We will be using a sequence from an isolate of SARS-CoV-2 sequenced early in the COVID-19 pandemic, in December 2019 in Wuhan, China. `data/comp90016_assignment_2_COVID19_N.fasta` is a FASTA file of the DNA sequence for the nucleocapsid (N) gene.\n",
    "\n",
    "`data/comp90016_assignment_2_9b.fasta` contains ORF-9b sequences from a selection of characterised coronaviruses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "dd6b43f21442b13e06883cc61a46eb9d",
     "grade": false,
     "grade_id": "cell-f53aec2e1fb97108",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "We will be using command line BLAST to execute a search.\n",
    "\n",
    "BLAST can be downloaded [here](https://www.ncbi.nlm.nih.gov/books/NBK279671/) OR installed using conda `conda install -c bioconda blast`\n",
    "\n",
    "Execute the following code block to complete the search. Note that `%%bash` signifies that the code block contains Linux commands and not Python code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "8df853f6f41711d7308f7a9b757a195f",
     "grade": false,
     "grade_id": "cell-68a8d06d95fb5527",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import Biopython to take advantage of Bio.Align objects for storing and accessing sequence alignments.\n",
    "from Bio import Align, Seq, SeqRecord\n",
    "# You may want to load the BLAST results with BlastIO https://biopython.org/docs/1.75/api/Bio.SearchIO.BlastIO.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# First let's check that BLAST is installed in the current environment and is available on your PATH\n",
    "which blastn\n",
    "# If the output is blank or you see an error, BLAST is not properly installed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "8977ab2e288d8729c6e07e4566e9b300",
     "grade": false,
     "grade_id": "cell-e537f8ff195f3be5",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Build a custom BLAST database from the ORF-9b sequences.\n",
    "echo \"Prepare blast database\"\n",
    "makeblastdb -dbtype nucl -in data/comp90016_assignment_2_9b.fasta\n",
    "\n",
    "# You will see some extra files have appeared.\n",
    "\n",
    "# Execute a BLAST search.\n",
    "echo \"Running blastn\"\n",
    "# The results will be returned in tabulated form and saved to a file.\n",
    "blastn -query data/comp90016_assignment_2_COVID19_N.fasta -db data/comp90016_assignment_2_9b.fasta -outfmt 7 -out blast_results.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Print contents of blastn output\n",
    "cat blast_results.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "be8f6c478cf06a105373cf6a3ce6ec7a",
     "grade": false,
     "grade_id": "cell-76e121e33e817785",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "### Questions\n",
    "In the cells below, complete the following tasks:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "2f388de555c119190e47dda078d2e5c7",
     "grade": false,
     "grade_id": "Q-3-1",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "\n",
    "<h3> Question 3.1 </h3>\n",
    "\n",
    "(5 marks, max 100 words)\n",
    "\n",
    "BLAST can be used to query a database for matches to a sequence of interest. Here we have created a custom database from a selection of coronavirus genomes, but we may also want to search for matches in a public database such as those available via NCBI, or in a custom database shared by another researcher.\n",
    "    \n",
    "\n",
    "Explain one factor that is important to consider when selecting an appropriate BLAST database.\n",
    "    \n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "The choice depends on the specific research question or objective at hand. It should contain sequences closely related to those we are querying or relevant to the biological context of our analysis. Factors to consider include the completeness of sequences (whole or fragments), level of annotation, available metadata, database size, frequency of updates for comprehensive and up to date information, and accessibility, as some databases may require specific access controls."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "130f7df2fcee2eeeab9654f29d2854d0",
     "grade": false,
     "grade_id": "Q-3-2",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "\n",
    "<h3> Question 3.2 </h3>\n",
    "\n",
    "(10 marks, max 100 words)\n",
    "\n",
    "Is there evidence from the BLAST results that the SARS-CoV-2 N gene has an internal ORF homologous to ORF-9b sequences in other coronaviruses? Justify your choice. What was the highest scoring hit?\n",
    "    \n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (10 marks) - complete this cell -- \n",
    "\n",
    "Yes, the results show multiple hits with high percentage of identities and statistically significant alignment scores (bit scores). The low E-values, approaching 0, indicate a strong confidence level in the matches and a minimal likelihood of the observed similarities occurring randomly. These findings strongly suggest that SARS-CoV-2 N gene and ORG-9b sequences are similar in other coronaviruses. The highest scoring hit achieved a score of 366, further proving the similarity. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "7e3bf0ed45fb68c837b117388058f5f2",
     "grade": false,
     "grade_id": "Q-3-3",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "\n",
    "<h3> Question 3.3 </h3>\n",
    "\n",
    "(5 marks, max 100 words)\n",
    "\n",
    "Interpret the E value of the top scoring BLAST hit. How confident are you in your BLAST results?\n",
    "    \n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "The E-value describes as the number of hits one can expect to see by chance when searching a database of a particular size. It decreases exponentially as the score of the match increases. The lower E-value the more significant the match is. In the context of the highest scoring BLAST, E-value 7.49e-105, it indicates a very high confidence level in the match. If we were to run the same search against a random database of a smiliar size, we would expect to find a match of this quality or better, less than 1x in 10^105 searches.\n",
    "Very confident."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "64e0da7365b7954672b4dcde79d0a5d0",
     "grade": false,
     "grade_id": "Q-3-4",
     "locked": true,
     "points": 5,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "\n",
    "<h3> Question 3.4 </h3>\n",
    "\n",
    "(5 marks, max 50 words)\n",
    "\n",
    "Modify the previous BLAST command to:\n",
    "\n",
    "- Work with an input of a protein sequence FASTA file (named `comp90016_assignment_2_COVID19_N_protein.fasta`). \n",
    "- Decrease the word-size parameter from the default value to 2. \n",
    "- Use the same custom database and output format used previously. \n",
    "    \n",
    "**Note that you are not required to execute this command.**\n",
    "    \n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (5 marks) - complete this cell -- \n",
    "\n",
    "blastp -query data/comp90016_assignment_2_COVID19_N_protein.fasta -db data/comp90016_assignment_2_9b.fasta -outfmt 7 -out blast_results.tab -word_size 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "6de8b97a44bfa4c7fc4526c2e081a480",
     "grade": false,
     "grade_id": "Q-3-5",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": true
    }
   },
   "source": [
    "<div class=\"question\">\n",
    "\n",
    "<h3> Question 3.5 </h3>\n",
    "\n",
    "(10 marks, max 100 words)\n",
    "\n",
    "What effect does decreasing the word size have on the sensitivity (ability to detect true homologs) and run-time of the search? Explain why this is the case.\n",
    "    \n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### -- GRADED CELL (10 marks) - complete this cell -- \n",
    "\n",
    "A BLAST search begins by identifying exact matches of the word-size between the query sequence and sequences in the database. Using smaller words increase the chances of discovering short regions of exact matches, enhancing the potential to find seed matches, and increasing sensitivity. Hence, increase the number of detected homologous sequences. However, this approach also leads to longer run-time as it involves processing a greater number of potential seed matches and performing more extensions and evaluations, consuming additional computational resources and time."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "87782ef3be7c9d7b5f4eb7f63519c449",
     "grade": false,
     "grade_id": "cell-d73ed6f61136388d",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "For the next question, we will be using amino-acid substitution matrices, encoded as 2D dictionaries. The cell below stores the BLOSUM-62 substitution matrix as a dictionary (the default substitution matrix for BLAST). The protein sequences used to test your code will only contain the one-letter codes of the 20 standard amino-acids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "3b97ce67723874b2eaed1c0e34c6abb5",
     "grade": false,
     "grade_id": "cell-84eff1f1350a2427",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "BLOSUM_62 = {\n",
    "    '*':{'*':1,'A':-4,'C':-4,'B':-4,'E':-4,'D':-4,'G':-4,'F':-4,'I':-4,'H':-4,'K':-4,'M':-4,'L':-4,'N':-4,'Q':-4,'P':-4,'S':-4,'R':-4,'T':-4,'W':-4,'V':-4,'Y':-4,'X':-4,'Z':-4},\n",
    "    'A':{'*':-4,'A':4,'C':0,'B':-2,'E':-1,'D':-2,'G':0,'F':-2,'I':-1,'H':-2,'K':-1,'M':-1,'L':-1,'N':-2,'Q':-1,'P':-1,'S':1,'R':-1,'T':0,'W':-3,'V':0,'Y':-2,'X':-1,'Z':-1},\n",
    "    'C':{'*':-4,'A':0,'C':9,'B':-3,'E':-4,'D':-3,'G':-3,'F':-2,'I':-1,'H':-3,'K':-3,'M':-1,'L':-1,'N':-3,'Q':-3,'P':-3,'S':-1,'R':-3,'T':-1,'W':-2,'V':-1,'Y':-2,'X':-1,'Z':-3},\n",
    "    'B':{'*':-4,'A':-2,'C':-3,'B':4,'E':1,'D':4,'G':-1,'F':-3,'I':-3,'H':0,'K':0,'M':-3,'L':-4,'N':3,'Q':0,'P':-2,'S':0,'R':-1,'T':-1,'W':-4,'V':-3,'Y':-3,'X':-1,'Z':1},\n",
    "    'E':{'*':-4,'A':-1,'C':-4,'B':1,'E':5,'D':2,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':-2,'L':-3,'N':0,'Q':2,'P':-1,'S':0,'R':0,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':4},\n",
    "    'D':{'*':-4,'A':-2,'C':-3,'B':4,'E':2,'D':6,'G':-1,'F':-3,'I':-3,'H':-1,'K':-1,'M':-3,'L':-4,'N':1,'Q':0,'P':-1,'S':0,'R':-2,'T':-1,'W':-4,'V':-3,'Y':-3,'X':-1,'Z':1},\n",
    "    'G':{'*':-4,'A':0,'C':-3,'B':-1,'E':-2,'D':-1,'G':6,'F':-3,'I':-4,'H':-2,'K':-2,'M':-3,'L':-4,'N':0,'Q':-2,'P':-2,'S':0,'R':-2,'T':-2,'W':-2,'V':-3,'Y':-3,'X':-1,'Z':-2},\n",
    "    'F':{'*':-4,'A':-2,'C':-2,'B':-3,'E':-3,'D':-3,'G':-3,'F':6,'I':0,'H':-1,'K':-3,'M':0,'L':0,'N':-3,'Q':-3,'P':-4,'S':-2,'R':-3,'T':-2,'W':1,'V':-1,'Y':3,'X':-1,'Z':-3},\n",
    "    'I':{'*':-4,'A':-1,'C':-1,'B':-3,'E':-3,'D':-3,'G':-4,'F':0,'I':4,'H':-3,'K':-3,'M':1,'L':2,'N':-3,'Q':-3,'P':-3,'S':-2,'R':-3,'T':-1,'W':-3,'V':3,'Y':-1,'X':-1,'Z':-3},\n",
    "    'H':{'*':-4,'A':-2,'C':-3,'B':0,'E':0,'D':-1,'G':-2,'F':-1,'I':-3,'H':8,'K':-1,'M':-2,'L':-3,'N':1,'Q':0,'P':-2,'S':-1,'R':0,'T':-2,'W':-2,'V':-3,'Y':2,'X':-1,'Z':0},\n",
    "    'K':{'*':-4,'A':-1,'C':-3,'B':0,'E':1,'D':-1,'G':-2,'F':-3,'I':-3,'H':-1,'K':5,'M':-1,'L':-2,'N':0,'Q':1,'P':-1,'S':0,'R':2,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':1},\n",
    "    'M':{'*':-4,'A':-1,'C':-1,'B':-3,'E':-2,'D':-3,'G':-3,'F':0,'I':1,'H':-2,'K':-1,'M':5,'L':2,'N':-2,'Q':0,'P':-2,'S':-1,'R':-1,'T':-1,'W':-1,'V':1,'Y':-1,'X':-1,'Z':-1},\n",
    "    'L':{'*':-4,'A':-1,'C':-1,'B':-4,'E':-3,'D':-4,'G':-4,'F':0,'I':2,'H':-3,'K':-2,'M':2,'L':4,'N':-3,'Q':-2,'P':-3,'S':-2,'R':-2,'T':-1,'W':-2,'V':1,'Y':-1,'X':-1,'Z':-3},\n",
    "    'N':{'*':-4,'A':-2,'C':-3,'B':3,'E':0,'D':1,'G':0,'F':-3,'I':-3,'H':1,'K':0,'M':-2,'L':-3,'N':6,'Q':0,'P':-2,'S':1,'R':0,'T':0,'W':-4,'V':-3,'Y':-2,'X':-1,'Z':0},\n",
    "    'Q':{'*':-4,'A':-1,'C':-3,'B':0,'E':2,'D':0,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':0,'L':-2,'N':0,'Q':5,'P':-1,'S':0,'R':1,'T':-1,'W':-2,'V':-2,'Y':-1,'X':-1,'Z':3},\n",
    "    'P':{'*':-4,'A':-1,'C':-3,'B':-2,'E':-1,'D':-1,'G':-2,'F':-4,'I':-3,'H':-2,'K':-1,'M':-2,'L':-3,'N':-2,'Q':-1,'P':7,'S':-1,'R':-2,'T':-1,'W':-4,'V':-2,'Y':-3,'X':-1,'Z':-1},\n",
    "    'S':{'*':-4,'A':1,'C':-1,'B':0,'E':0,'D':0,'G':0,'F':-2,'I':-2,'H':-1,'K':0,'M':-1,'L':-2,'N':1,'Q':0,'P':-1,'S':4,'R':-1,'T':1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':0},\n",
    "    'R':{'*':-4,'A':-1,'C':-3,'B':-1,'E':0,'D':-2,'G':-2,'F':-3,'I':-3,'H':0,'K':2,'M':-1,'L':-2,'N':0,'Q':1,'P':-2,'S':-1,'R':5,'T':-1,'W':-3,'V':-3,'Y':-2,'X':-1,'Z':0},\n",
    "    'T':{'*':-4,'A':0,'C':-1,'B':-1,'E':-1,'D':-1,'G':-2,'F':-2,'I':-1,'H':-2,'K':-1,'M':-1,'L':-1,'N':0,'Q':-1,'P':-1,'S':1,'R':-1,'T':5,'W':-2,'V':0,'Y':-2,'X':-1,'Z':-1},\n",
    "    'W':{'*':-4,'A':-3,'C':-2,'B':-4,'E':-3,'D':-4,'G':-2,'F':1,'I':-3,'H':-2,'K':-3,'M':-1,'L':-2,'N':-4,'Q':-2,'P':-4,'S':-3,'R':-3,'T':-2,'W':11,'V':-3,'Y':2,'X':-1,'Z':-3},\n",
    "    'V':{'*':-4,'A':0,'C':-1,'B':-3,'E':-2,'D':-3,'G':-3,'F':-1,'I':3,'H':-3,'K':-2,'M':1,'L':1,'N':-3,'Q':-2,'P':-2,'S':-2,'R':-3,'T':0,'W':-3,'V':4,'Y':-1,'X':-1,'Z':-2},\n",
    "    'Y':{'*':-4,'A':-2,'C':-2,'B':-3,'E':-2,'D':-3,'G':-3,'F':3,'I':-1,'H':2,'K':-2,'M':-1,'L':-1,'N':-2,'Q':-1,'P':-3,'S':-2,'R':-2,'T':-2,'W':2,'V':-1,'Y':7,'X':-1,'Z':-2},\n",
    "    'X':{'*':-4,'A':-1,'C':-1,'B':-1,'E':-1,'D':-1,'G':-1,'F':-1,'I':-1,'H':-1,'K':-1,'M':-1,'L':-1,'N':-1,'Q':-1,'P':-1,'S':-1,'R':-1,'T':-1,'W':-1,'V':-1,'Y':-1,'X':-1,'Z':-1},\n",
    "    'Z':{'*':-4,'A':-1,'C':-3,'B':1,'E':4,'D':1,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':-1,'L':-3,'N':0,'Q':3,'P':-1,'S':0,'R':0,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':4}}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "2bf9aa67de0bb3883f756b9d4bd17a49",
     "grade": false,
     "grade_id": "Q-3-6-info",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "<div class=\"info\">\n",
    "<h3> Question 3.6 </h3>\n",
    "\n",
    "(10 marks)\n",
    "\n",
    "<b>Challenge:</b> Write a Python function to calculate the percentage sequence similarity between two aligned protein sequences. The percentage similarity is defined as the number of positions with similar amino acid residues as a percentage of the total number of positions in the alignment. \n",
    "    \n",
    "Two amino acid residues are considered similar based on the input substitution matrix. If two amino acids have a value greater than 0 in their corresponding cell of the input matrix, they are similar. \n",
    "\n",
    "- [ ] Input: \n",
    "    - Pairwise_aln as a `Bio.Align.MultipleSeqAlignment` object that contains only two aligned sequences.\n",
    "    - sub_matrix as a 2D dictionary encoding an amino-acid substitution matrix, such as the BLOSUM-62 matrix above.\n",
    "- [ ] Output: Return a floating-point number between 0 and 100. \n",
    "- [ ] Round the output to 4 decimal places.\n",
    "- [ ] If pairwise_aln does not contain 2 sequences, return None.\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "8dee8d238484e9ac171b3991fb26ceaf",
     "grade": false,
     "grade_id": "Q-3-6",
     "locked": false,
     "schema_version": 3,
     "solution": true,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# GRADED CELL 3.6 (10 marks, max 1 min run-time)\n",
    "%time\n",
    "def percentage_similarity(pairwise_aln, sub_matrix):\n",
    "    \"\"\"\n",
    "    Calculate the percentage sequence similarity between two aligned protein sequences. \n",
    "    Assume that pairwise_aln is an Bio.Align.MultipleSeqAlignment object. \n",
    "    Assume sub_matrix is a 2D dictionary encoding an amino-acid substitution matrix. \n",
    "    Two amino acid residues are similar if there is a value greater than 0 in the corresponding cell of sub_matrix. \n",
    "    The percentage similarity is defined as the number of positions with similar amino acid residues as a percentage of the total number of positions in the alignment.\n",
    "    Return a floating point number between 0 and 100.\n",
    "    Round the output to 4 decimal places.\n",
    "    If pairwise_aln does not contain 2 sequences, return None.\n",
    "    \"\"\"\n",
    "   \n",
    "    # YOUR CODE HERE\n",
    "    similar_aa = 0\n",
    "    \n",
    "    if len(pairwise_aln) == 2:\n",
    "        seq1 = pairwise_aln[0].seq\n",
    "        seq2 = pairwise_aln[1].seq\n",
    "\n",
    "        for s1,s2 in zip(seq1,seq2):\n",
    "            if s1 != '-' and s2 != '-' and sub_matrix[s1][s2] > 0:\n",
    "                similar_aa += 1\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "        similarity = similar_aa / len(seq1) * 100\n",
    "        \n",
    "    else:\n",
    "        return None\n",
    "\n",
    "    return round(similarity, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ~~ Test your function in this cell ~~\n",
    "demo_sequence_a = SeqRecord.SeqRecord(Seq.Seq(\"MYWIW\"))\n",
    "demo_sequence_b = SeqRecord.SeqRecord(Seq.Seq(\"IYW--\"))\n",
    "demo_pairwise_prot_aln = Align.MultipleSeqAlignment([demo_sequence_a, demo_sequence_b])\n",
    "\n",
    "sequence_a = SeqRecord.SeqRecord(Seq.Seq(\"MYGEGEPGGWQDHVTVLATRRHPKWAQAWVSTMPWGYECGFSRAWVHQTPWINV-----VSLSSHEAYGVVAVRHPWEIFSPYEVYAPYVQDTQHHGNPGQFTTSCYPDE\"))\n",
    "sequence_b = SeqRecord.SeqRecord(Seq.Seq(\"MYADGEPGAWQDHMTVLAIYWHHKWAHAWVSTMPWSYECGFSRAWVHQTPWINVIRFTQVSLSSRAWYGILAVRHPWEIFSPYDVYAPYVAATQHHGNPGQFSTSCYP--\"))\n",
    "pairwise_prot_aln = Align.MultipleSeqAlignment([sequence_a, sequence_b])\n",
    "\n",
    "print(percentage_similarity(demo_pairwise_prot_aln, BLOSUM_62)) # should return 60.0\n",
    "\n",
    "print(percentage_similarity(pairwise_prot_aln, BLOSUM_62))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "0718aa3d4cc3948b3f92a48fff269474",
     "grade": true,
     "grade_id": "T-3-6a",
     "locked": true,
     "points": 10,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "outputs": [],
   "source": [
    "# --- AUTOGRADING CELL DO NOT EDIT ----\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "26db4d8c34c57ba5821a9d2cc35e1c53",
     "grade": false,
     "grade_id": "cell-5c804f1187504fa9",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "# END OF ASSIGNMENT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "a63655b02b8d755e8712ad29d6217223",
     "grade": false,
     "grade_id": "cell-c82c42eca2b5bee4",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "## Submitting\n",
    "\n",
    "Follow these steps to submit your assignment\n",
    "\n",
    "1) Before you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\\rightarrow$Run All).\n",
    "\n",
    "2) Make sure you have filled in any place that says `YOUR CODE HERE` or \"YOUR ANSWER HERE\"\n",
    "\n",
    "3) **Clear all output** (in the menubar, select Kernel$\\rightarrow$Restart & Clear Output)\n",
    "\n",
    "4) Your completed notebook file containing all your answers must be turned in via Canvas in `.ipynb` format.\n",
    "\n",
    "5) You must also submit a copy of this notebook in `html` format with the output cleared.\n",
    "\n",
    "\n",
    "Your submission should include **only two** files with names formatted as: **Assignment_2.ipynb** and **Assignment_2.html**\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
